### PR TITLE
Tests for text gen output text

### PR DESCRIPTION
**Pull request by vidyasiv**
# What does this PR do?

Fixes # (issue)
- Fix to `run_generation.py` so `results.json` stores all outputs instead of just the last when batch_size > 1
- Functional verification of select text gen models to be run for every PR (fast tests)
- Duration of added fast tests on Gaudi2: ` 7 passed, 1 warning in 1991.08s (0:33:11)`
- Functional verification added to slow test `test_text_generation_example.py` for select BF16 1HPU cases
- Found issues in bigcode/starcoder output in fast(passes) vs slow(fails due to bucketing options), will file ticket to fix that.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by vidyasiv** [comment]
@regisss , these tests add 30min to fast tests, let me know if you'd like me to cut down on the models

**Comment by jiminha** [comment]
@vidyasiv  this is good, but is it possible we actually add this accuracy check also as part of the test_text_generation_example.py? We already run so many models test there for token generation, so I think it will not add much time to just compare the 1st or last token outputs along with checking throughputs.

**Comment by vidyasiv** [comment]
> @vidyasiv this is good, but is it possible we actually add this accuracy check also as part of the test_text_generation_example.py? We already run so many models test there for token generation, so I think it will not add much time to just compare the 1st or last token outputs along with checking throughputs.

It won't be part of every PR check then i.e fast tests.

**Comment by jiminha** [comment]
> > @vidyasiv this is good, but is it possible we actually add this accuracy check also as part of the test_text_generation_example.py? We already run so many models test there for token generation, so I think it will not add much time to just compare the 1st or last token outputs along with checking throughputs.
> 
> It won't be part of every PR check then i.e fast tests.

Got it. Then, can we keep this as it is for fast test, but also add similar logic for the existing text-gen test?

File: tests/test_text_generation_example.py [hunk 1]
```diff
@@ -19,32 +21,38 @@
...
+            ("bigcode/starcoder2-3b", 1, False, 261.07213776344133, True),
+            ("adept/persimmon-8b-base", 4, False, 366.73968820698406, False),
+            ("Qwen/Qwen1.5-7B", 4, False, 490.8621617893209, False),
+            ("google/gemma-7b", 1, False, 109.70751574382221, True),
+            ("state-spaces/mamba-130m-hf", 1536, False, 5385.511100161605, False),
+            ("Deci/DeciLM-7B", 1, False, 120, False),
+            ("Qwen/Qwen2-7B", 512, False, 9669.45787, True),
+            ("Qwen/Qwen1.5-MoE-A2.7B", 1, True, 44.25834541569395, False),
+            ("EleutherAI/gpt-neo-2.7B", 1, False, 257.2476416844122, False),
         ],
```
Comment (jiminha) [inline tests/test_text_generation_example.py:56]: So only the model with "True" will have output tested? If we want to add more output check later on, all we need to do is change this last value to True, and add the values 1st token output to the "MODEL_OUTPUTS" table below?

File: tests/test_text_generation_example.py [hunk 2]
```diff
@@ -89,41 +97,51 @@
             ("gpt2-xl", 1, False, 51.61471298016438),
         ],
     }
+    MODEL_OUTPUTS = {
```
Comment (jiminha) [inline tests/test_text_generation_example.py:100]: Can this be also used for fp8 test?

File: tests/test_text_generation_example.py (hunk 1 above)
Comment (vidyasiv) [reply tests/test_text_generation_example.py:56]: yes that is right

File: tests/test_text_generation_example.py (hunk 2 above)
Comment (vidyasiv) [reply tests/test_text_generation_example.py:100]: i have not tried but perhaps it can be a later effort

Comment (jiminha) [reply tests/test_text_generation_example.py:100]: Yeah I think we should add it for fp8 output check as well - with fp8, and int4. Maybe for another PR.

**Comment by jiminha** [comment]
@vidyasiv  Is this time including model downloading time? 
Duration of added fast tests on Gaudi2:  7 passed, 1 warning in 1991.08s (0:33:11)

**Comment by vidyasiv** [comment]
@jiminha , i deleted cache and kicked off a run, will let u know.

File: tests/test_functional_text_generation_example.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        "bigcode/starcoder2-3b": 'def print_hello_world():\n    print("Hello World")\n\ndef print_hello_world_with_name(name):\n    print("Hello World, " + name)\n\ndef print_hello_world_with_name_and_age(name, age):\n    print("Hello World, " + name + ", " + str(age))\n\ndef print_hello_world_with_name_and_age_and_gender(name, age, gender):\n    print("Hello',
+        "google/gemma-7b": "DeepSpeed is a machine learning framework that enables training of large-scale models on commodity hardware. It is designed to be a drop-in replacement for PyTorch, and it is compatible with the existing PyTorch ecosystem. DeepSpeed is designed to be easy to use, and it provides a number of features that make it easy to train large-scale models.\n\nDeepSpeed is a machine learning framework that enables training of large-scale models on commodity hardware. It is designed to be a drop-in replacement for PyTorch, and",
```
Comment (vidyasiv) [inline tests/test_functional_text_generation_example.py:17]: @regisss , a common theme I observe in the output (except starcoder/starcoder2) is repeated sentences, not sure if that is some other bug or expected

**Comment by vidyasiv** [comment]
@jiminha  can u approve/ apply run-test?

**Comment by jiminha** [comment]
@regisss  Could you check this PR?  This will run basic accuracy check on the fast test (48min added) to PR. 
Is this ok?

**Review by jiminha — APPROVED** [review]

File: tests/test_functional_text_generation_example.py (hunk 3 above)
Comment (regisss) [reply tests/test_functional_text_generation_example.py:17]: I'm not surprised, greedy search tends to produce very repetitive outputs as it only looks for the most likely token to generate in each iteration. Tweaking some generation parameters (for example to penalize more the sequences of tokens that are already part of the output) and/or using sampling (to consider several candidates and not only the most likely one) usually helps to give the model more flexibility and to get more realistic outputs.

File: Makefile [hunk 4]
```diff
@@ -35,6 +35,7 @@ style: clean
 fast_tests:
 	python -m pip install .[tests]
 	python -m pytest tests/test_gaudi_configuration.py tests/test_trainer_distributed.py tests/test_trainer.py tests/test_trainer_seq2seq.py
+	python -m pytest test_functional_text_generation_example.py
```
Comment (regisss) [inline Makefile:38]: I think it's a bit tricky to add this one to the fast tests now as it's going to be quite slow with the current infrastructure. Maybe we can comment this line and add a TODO comment to uncomment it when we have more servers dedicated to CI.

**Review by regisss — COMMENTED** [review]
Very nice PR!

I would just not add this type of check to the fast tests yet, see my comment below. The reason is that I recently migrated the fast tests from DL1 instances to Gaudi2. And I have only one Gaudi2 server for the whole CI (i.e. all the fast tests of all the PRs + nightly tests) vs up to 4 DL1 instances in parallel before. I'm afraid this would increase the queue of workflows quite a lot. Happy to reconsider it if/when I have at least one more server for CI. In the meantime, these checks will still be performed in the nightly tests.

File: Makefile (hunk 4 above)
Comment (vidyasiv) [reply Makefile]: updated

**Review by regisss — APPROVED** [review]

**Pull request merged**